bdpy/dl/torch/torch.py

```diff
@@ -17,128 +17,108 @@
-            layer_object.register_forward_hook(self._extractor)
+            Whether to detach the extracted feature tensors. Defaults to False.
+        transform :callable, optional 
+            A function that receives extracted features dict and returns transformed features.
+            Example usage: turning a tuple into a tensor on transformer models.
+        """
+        self.encoder = encoder
+        self.layers = layers
+        self.layer_mapping = layer_mapping if layer_mapping else {layer: layer for layer in layers}
+        self.device = device
+        self.detach = detach
+        self.transform = transform
 
-    def __call__(self, x: _tensor_t) -> Dict[str, np.ndarray] | Dict[str, torch.Tensor]:
+        self._features = OrderedDict()  # dict to store extracted features
+        self._hook_handler = {}         # layer name -> hook handler object for deleting them
+        self._register_hooks()
+
+        self.encoder.to(self.device)
+
+    def _register_hooks(self):
+        """Registers forward hooks to extract features from specified layers."""
+        for name in self.layers:
+            mapped_name = self.layer_mapping[name]
+            layer_obj = models._parse_layer_name(self.encoder, mapped_name)
+            hook_handle = layer_obj.register_forward_hook(
+                _Hook(name, self._features, self.detach)
+            )
+            self._hook_handler[name] = hook_handle  # Store the actual hook handle
+    
+    def __call__(self, x: Union[torch.Tensor, np.ndarray]) -> Dict[str, Union[torch.Tensor, np.ndarray]]:
         return self.run(x)
-
-    def run(self, x: _tensor_t) -> Dict[str, np.ndarray] | Dict[str, torch.Tensor]:
-        '''Extract feature activations from the specified layers.
-
+    
+    def run(self, x: Union[torch.Tensor, np.ndarray]) -> Dict[str, Union[torch.Tensor, np.ndarray]]:
+        """
+        Feeds input through the encoder and extracts features.
+        
         Parameters
         ----------
-        x : numpy.ndarray or torch.Tensor
-            Input image (numpy.ndarray or torch.Tensor).
-
-        Returns
-        -------
-        features : Dict[str, Union[numpy.ndarray, torch.Tensor]]
-            Feature activations from the specified layers.
-            Each key is the layer name and each value is the feature activation.
-        '''
-
-        self._extractor.clear()
+        x : torch.Tensor or np.ndarray
+            Input tensor to be processed by the encoder.
+        
+        Returns:
+        ----------
+        dict[str, Union[torch.Tensor, np.ndarray]]
+            A dictionary containing features extracted from the specified layers, 
+            optionally transformed.
+        """
         if not isinstance(x, torch.Tensor):
-            xt = torch.tensor(x[np.newaxis], device=self.__device)
-        else:
-            xt = x
+            # TODO: From legacy code. Is this necessary?
+            x = torch.tensor(x[np.newaxis], device=self.device)
```

Comment:
```suggestion
            warnings.warn(
                "In future versions, it will no longer be possible to input ndarrays " \
                "directly into FeatureExtractor. Please convert the input to torch.Tensor.",
                DeprecationWarning,
                stacklevel=2
            )
            x = torch.tensor(x[np.newaxis], device=self.device)
```